rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -144,25 +146,19 @@ NodeParameters::NodeParameters(
 
     // Combine parameter yaml files, overwriting values in older ones
     for (auto & param : iter->second) {
-      parameters[param.get_name()] = param;
+      initial_parameter_values_[param.get_name()] =
+        rclcpp::ParameterValue(param.get_value_message());
     }
   }
 
   // initial values passed to constructor overwrite yaml file sources
   for (auto & param : initial_parameters) {
-    parameters[param.get_name()] = param;
-  }
-
-  std::vector<rclcpp::Parameter> combined_values;
-  combined_values.reserve(parameters.size());
-  for (auto & kv : parameters) {
-    combined_values.emplace_back(kv.second);
+    initial_parameter_values_[param.get_name()] =
+      rclcpp::ParameterValue(param.get_value_message());
   }
 
-  // TODO(sloretz) store initial values and use them when a parameter is created ros2/rclcpp#475
-  // Set initial parameter values
-  if (!combined_values.empty()) {
-    rcl_interfaces::msg::SetParametersResult result = set_parameters_atomically(combined_values);
+  if (!initial_parameters.empty() && allow_undeclared_) {
```

Comment:
In the case where `!initial_parameters.empty()` and `!allow_undeclared_`, it seems like we are going to silently ignore the values coming from YAML.  Is that intentional?  Should we throw an exception instead?

Reply:
It is intentional, I have a follow up pr where the user can choose what to do with "extra" parameters in the YAML which are not declared in the source code. As well as they can run a checker for unused YAML parameters (i.e. raise an error if any extra YAML parameters are unused). The latter cannot be done automatically because we don't know when the user will be done declaring parameters, so they have to invoke the "checker".

This branch has (some of) my WIP:

https://github.com/ros2/rclcpp/commit/36ca81381355247ac03c3b2f0bd787422655a91e#diff-b2109335f72aba8ba2975b698da15ebd